File: main/java/com/websarva/wings/android/dowasurememo/WishlistMemo.java
package com.websarva.wings.android.dowasurememo;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;

public class WishlistMemo extends AppCompatActivity {

    private Databasehelper _helper;
    private String _category = "wishlist";

    private int indexCounter=2;
    int tagId;
    String table="wishlist";
    Context context=WishlistMemo.this;

    LayoutInflater inflater;
    LinearLayout llWishlistLayout;
    LinearLayout llWishlistInputform;
    LinearLayout llWishlistTitle;

    EditText etWishlistTitle;
    ImageButton btDelete;

    String strWishlistTitle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_wishlist_memo);

        Intent intent = getIntent();
        llWishlistLayout = findViewById(R.id.ll_wishlist_layout);

//        llWishlistLayout.removeAllViews();
//        DatabaseControl control4=new DatabaseControl(context,table);
//        control4.DatabaseAllDelete();

        _helper=new Databasehelper(getApplicationContext());
        SQLiteDatabase db=_helper.getWritableDatabase();
        String sqlSelect="SELECT * FROM wishlist";
        Cursor cursor=db.rawQuery(sqlSelect,null);
        cursor.moveToFirst();
        while (cursor.moveToNext()) {
            int i = cursor.getColumnIndex("_id");
            tagId = cursor.getInt(i);

            inflater = LayoutInflater.from(getApplicationContext());
            llWishlistLayout = findViewById(R.id.ll_wishlist_layout);
            llWishlistInputform=(LinearLayout)inflater.inflate(R.layout.wishlist_inputform,null);
            llWishlistLayout.addView(llWishlistInputform);

            llWishlistTitle=llWishlistInputform.findViewById(R.id.ll_wishlist_title);

            etWishlistTitle=llWishlistTitle.findViewById(R.id.et_wishlist_title);
            btDelete=llWishlistTitle.findViewById(R.id.bt_delete);
            btDelete.setOnClickListener
                    (new DeleteButton(WishlistMemo.this,llWishlistLayout,llWishlistInputform,table));

            i = cursor.getColumnIndex("wishlisttitle");
            strWishlistTitle = cursor.getString(i);

            try {
                etWishlistTitle.setText(strWishlistTitle);
            } catch (NullPointerException e) {
                strWishlistTitle = "";
            }

        }

        if(llWishlistLayout.getChildCount()!=0){
            LinearLayout firstView= (LinearLayout) llWishlistLayout.getChildAt(0);
            firstView.setVisibility(View.GONE);
        }else {
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater=getMenuInflater();
        inflater.inflate(R.menu.optionmenu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.option_add:
                if(llWishlistLayout.getChildCount()==0){
                    inflater = LayoutInflater.from(getApplicationContext());
                    llWishlistLayout = findViewById(R.id.ll_wishlist_layout);
                    llWishlistInputform=(LinearLayout)inflater.inflate(R.layout.wishlist_inputform,null);
                    llWishlistLayout.addView(llWishlistInputform);
                    llWishlistInputform.setVisibility(View.GONE);

                    String str="";
                    DatabaseControl control = new DatabaseControl(context, table);
                    control.deleteDatabase(1);

                    String column1="wishlisttitle";

                    DatabaseControl control2 = new DatabaseControl
                            (context, table,1, _category, str);
                    control2.insertDatabaseOneColumns(column1);
                }
                inflater = LayoutInflater.from(getApplicationContext());
                llWishlistInputform=(LinearLayout)inflater.inflate(R.layout.wishlist_inputform,null);
                llWishlistLayout.addView(llWishlistInputform);

                llWishlistTitle=llWishlistInputform.findViewById(R.id.ll_wishlist_title);

                etWishlistTitle=llWishlistTitle.findViewById(R.id.et_wishlist_title);
                btDelete=llWishlistTitle.findViewById(R.id.bt_delete);
                btDelete.setOnClickListener
                        (new DeleteButton(WishlistMemo.this,llWishlistLayout,llWishlistInputform,table));
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPause() {
        super.onPause();

        indexCounter = 2;

        for (int i = 0; i < llWishlistLayout.getChildCount(); i++) {
            LinearLayout linearLayout = (LinearLayout) llWishlistLayout.getChildAt(i);
            etWishlistTitle = linearLayout.findViewById(R.id.et_wishlist_title);

            strWishlistTitle = etWishlistTitle.getText().toString();


            DatabaseControl control = new DatabaseControl(context, table);
            control.deleteDatabase(indexCounter);

            String column1="wishlisttitle";


            DatabaseControl control2=new DatabaseControl
                    (context,table,indexCounter,_category,strWishlistTitle);
            control2.insertDatabaseOneColumns(column1);

            Log.d("pause358", "" + indexCounter);
            indexCounter++;

        }
    }
}

File: main/java/com/websarva/wings/android/dowasurememo/WishlistMemo.kt
package com.websarva.wings.android.dowasurememo

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import android.widget.ImageButton
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity

/**
 * 目標をメモするWishlistメモのクラス
 *
 * @author nakayama
 * @version 1.0
 */
class WishlistMemo : AppCompatActivity() {
    private val context: Context = this@WishlistMemo
    private var inflater: LayoutInflater? = null
    private var llWishlistLayout: LinearLayout? = null
    private var llWishlistInputform: LinearLayout? = null
    private var llWishlistTitle: LinearLayout? = null
    private var etWishlistTitle: EditText? = null
    private var btDelete: ImageButton? = null
    private var strWishlistTitle: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_wishlist_memo)
        inflater = LayoutInflater.from(applicationContext)
        llWishlistLayout = findViewById(R.id.ll_wishlist_layout)
        llWishlistInputform = inflater.inflate(R.layout.wishlist_inputform, null) as LinearLayout
        /**
         * データベースの列名の配列
         */
        val columnNames = listOf("wishlisttitle")

        //データベースからデータを取り出して、レイアウトを作成する処理
        val control = DatabaseControl(context, TABLE, columnNames)
        control.selectDatabase(llWishlistLayout)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.optionmenu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.option_add -> {
                //オプションメニューの＋ボタンを押すと、動的にビューを追加する処理
                inflater = LayoutInflater.from(applicationContext)
                llWishlistInputform = inflater.inflate(R.layout.wishlist_inputform, null) as LinearLayout
                llWishlistLayout!!.addView(llWishlistInputform)
                llWishlistTitle = llWishlistInputform!!.findViewById(R.id.ll_wishlist_title)
                etWishlistTitle = llWishlistTitle.findViewById(R.id.et_wishlist_title)
                btDelete = llWishlistTitle.findViewById(R.id.bt_delete)
                btDelete.setOnClickListener(DeleteButton(this@WishlistMemo, llWishlistLayout, llWishlistInputform, TABLE))
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        //データベースにある全てのデータを削除
        val control = DatabaseControl(context, TABLE)
        control.deleteAllDatabase()

        //メモの文字列を取得してデータベースにインサートする
        for (i in 0 until llWishlistLayout!!.childCount) {
            val linearLayout = llWishlistLayout!!.getChildAt(i) as LinearLayout
            etWishlistTitle = linearLayout.findViewById(R.id.et_wishlist_title)
            strWishlistTitle = etWishlistTitle.getText().toString()
            val control2 = DatabaseControl(context, TABLE, i, _CATEGORY, strWishlistTitle)
            control2.insertDatabaseOneColumns("wishlisttitle")
        }
        finish()
    }

    companion object {
        private const val _CATEGORY = "WISHLIST"

        /**
         * データベースのテーブル名
         */
        private const val TABLE = "wishlist"
    }
}